nomos-mix/network/src/behaviour.rs

```diff
@@ -58,33 +59,41 @@ impl Behaviour {
         }
     }
 
-    /// Publishs a message through the mix network.
-    ///
-    /// This function expects that the message was already encoded for the cryptographic mixing
-    /// (e.g. Sphinx encoding).
-    ///
-    /// The message is forward to all connected peers,
-    /// so that it can arrive in the mix node who can unwrap it one layer.
-    /// Fully unwrapped messages are returned as the [`MixBehaviourEvent::FullyUnwrappedMessage`].
+    /// Publish a message (data or drop) to all connected peers
     pub fn publish(&mut self, message: Vec<u8>) -> Result<(), Error> {
-        self.duplicate_cache.cache_set(message_id(&message), ());
-        self.forward_message(message, None)
+        if is_drop_message(&message) {
+            // Bypass deduplication for the drop message
+            return self.forward_message(message, None);
+        }
+
+        let msg_id = message_id(&message);
+        // If the message was already seen, don't forward it again
+        if self.duplicate_cache.cache_get(&msg_id).is_some() {
+            return Ok(());
+        }
+
+        let result = self.forward_message(message, None);
+        // Add the message to the cache only if the forwarding was successfully triggered
+        if result.is_ok() {
+            self.duplicate_cache.cache_set(msg_id, ());
+        }
+        result
     }
 
-    /// Forwards a message to all connected peers except the one that was received from.
+    /// Forwards a message to all connected peers except the excluded peer.
     ///
     /// Returns [`Error::NoPeers`] if there are no connected peers that support the mix protocol.
     fn forward_message(
         &mut self,
         message: Vec<u8>,
-        propagation_source: Option<PeerId>,
+        excluded_peer: Option<PeerId>,
     ) -> Result<(), Error> {
         let peer_ids = self
             .negotiated_peers
             .keys()
             .filter(|&peer_id| {
-                if let Some(propagation_source) = propagation_source {
-                    *peer_id != propagation_source
+                if let Some(excluded_peer) = excluded_peer {
+                    *peer_id != excluded_peer
                 } else {
                     true
                 }
```

Comment:
Since we collect. We could collect into a hashset directly. Then just remove the excluded peer.

```suggestion
let mut peer_ids: HashSet<_> = self
            .negotiated_peers
            .keys()
            .collect();
 if let Some(peer) = excluded_peer {
     peer_ids.remove(peer);
 }
```

Reply:
Much cleaner. Thank you! 01503d7d29cbe32ea790951ab0c8dfc40a202ac0